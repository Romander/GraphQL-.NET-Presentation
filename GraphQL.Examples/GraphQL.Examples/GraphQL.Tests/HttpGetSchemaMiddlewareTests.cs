using GraphQL.Tests.Utilities;
using HotChocolate.AspNetCore;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.TestHost;
using Snapshooter.Xunit;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using Xunit;

namespace GraphQL.Tests
{
  public class HttpGetSchemaMiddlewareTests : ServerTestBase
  {
    public HttpGetSchemaMiddlewareTests(TestServerFactory serverFactory)
        : base(serverFactory)
    {
    }

    [Fact]
    public async Task Download_GraphQL_SDL()
    {
      // arrange
      TestServer server = CreateStarWarsServer();
      var url = TestServerExtensions.CreateUrl("/graphql?sdl");
      var request = new HttpRequestMessage(HttpMethod.Get, url);

      // act
      HttpResponseMessage response = await server.CreateClient().SendAsync(request);

      // assert
      Assert.Equal(HttpStatusCode.OK, response.StatusCode);
      var result = await response.Content.ReadAsStringAsync();
      result.MatchSnapshot();
    }

    [Fact]
    public async Task Download_GraphQL_SDL_Disabled()
    {
      // arrange
      TestServer server = CreateStarWarsServer(
          configureConventions: e => e.WithOptions(
              new GraphQLServerOptions
              {
                EnableSchemaRequests = false
              }));
      var url = TestServerExtensions.CreateUrl("/graphql?sdl");
      var request = new HttpRequestMessage(HttpMethod.Get, url);

      // act
      HttpResponseMessage response = await server.CreateClient().SendAsync(request);

      // assert
      Assert.Equal(HttpStatusCode.BadRequest, response.StatusCode);
      var result = await response.Content.ReadAsStringAsync();
      result.MatchSnapshot();
    }
  }
}
