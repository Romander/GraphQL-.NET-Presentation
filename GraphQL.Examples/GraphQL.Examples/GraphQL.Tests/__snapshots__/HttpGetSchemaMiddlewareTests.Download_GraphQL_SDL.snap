schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}

"A character in the Star Wars universe."
interface Character {
  "The unique identifier for the character."
  id: ID!
  "The name of the character."
  name: String
  "The names of the character's friends."
  friends(first: Int after: String last: Int before: String): CharacterConnection
  "The episodes the character appears in."
  appearsIn: [Episode]
  "The height of the character."
  height(unit: Unit): Float
}

"A connection to a list of items."
type CharacterConnection {
  "Information to aid in pagination."
  pageInfo: PageInfo!
  "A list of edges."
  edges: [CharacterEdge!]
  "A flattened list of the nodes."
  nodes: [Character]
}

"An edge in a connection."
type CharacterEdge {
  "A cursor for use in pagination."
  cursor: String!
  "The item at the end of the edge."
  node: Character
}

"A droid in the Star Wars universe."
type Droid implements Character {
  "The unique identifier for the character."
  id: ID!
  "The episodes the character appears in."
  appearsIn: [Episode]
  friends(first: Int after: String last: Int before: String): CharacterConnection
  height(unit: Unit): Float
  "The name of the character."
  name: String
  "The droid's primary function."
  primaryFunction: String
}

"A human character in the Star Wars universe."
type Human implements Character {
  "The unique identifier for the character."
  id: ID!
  "The episodes the character appears in."
  appearsIn: [Episode]
  friends(first: Int after: String last: Int before: String): CharacterConnection
  height(unit: Unit): Float
  "The name of the character."
  name: String
  "The planet the character is originally from."
  homePlanet: String
}

type Mutation {
  "Creates a review for a given Star Wars episode.\n\n\n**Returns:**\nThe created review."
  createReview("The episode to review." episode: Episode! "The review." review: ReviewInput!): Review!
}

"Information about pagination in a connection."
type PageInfo {
  "Indicates whether more edges exist following the set defined by the clients arguments."
  hasNextPage: Boolean!
  "Indicates whether more edges exist prior the set defined by the clients arguments."
  hasPreviousPage: Boolean!
  "When paginating backwards, the cursor to continue."
  startCursor: String
  "When paginating forwards, the cursor to continue."
  endCursor: String
}

type Query {
  "Retrieve a hero by a particular Star Wars episode.\n\n\n**Returns:**\nThe character."
  hero("The episode to look up by." episode: Episode! = NEW_HOPE): Character
  character(characterIds: [String]): [Character!]!
  search(text: String): [SearchResult]
  "Gets a human by Id.\n\n\n**Returns:**\nThe human."
  human("The Id of the human to retrieve." id: String): Human
  "Get a particular droid by Id.\n\n\n**Returns:**\nThe droid."
  droid("The Id of the droid." id: String): Droid
}

"A review of a particular movie."
type Review {
  "The number of stars given for this review."
  stars: Int!
  "An explanation for the rating."
  commentary: String
}

"A starship in the Star Wars universe."
type Starship {
  "The Id of the starship."
  id: String
  "The name of the starship."
  name: String
  "The length of the starship."
  length: Float!
}

type Subscription {
  onReview(episode: Episode!): Review!
}

union SearchResult = Starship | Human | Droid

"A review of a particular movie."
input ReviewInput {
  "The number of stars given for this review."
  stars: Int!
  "An explanation for the rating."
  commentary: String
}

"The Star Wars episodes."
enum Episode {
  "Star Wars Episode IV: A New Hope"
  NEW_HOPE
  "Star Wars Episode V: Empire Strikes Back"
  EMPIRE
  "Star Wars Episode VI: Return of the Jedi"
  JEDI
}

"Different units of measurement."
enum Unit {
  FOOT
  METERS
}

"The `@defer` directive may be provided for fragment spreads and inline fragments to inform the executor to delay the execution of the current fragment to indicate deprioritization of the current fragment. A query with `@defer` directive will cause the request to potentially return multiple responses, where non-deferred data is delivered in the initial response and data deferred is delivered in a subsequent response. `@include` and `@skip` take precedence over `@defer`."
directive @defer("If this argument label has a value other than null, it will be passed on to the result of this defer directive. This label is intended to give client applications a way to identify to which fragment a deferred result belongs to." label: String "Deferred when true." if: Boolean) on FRAGMENT_SPREAD | INLINE_FRAGMENT

"The @deprecated directive is used within the type system definition language to indicate deprecated portions of a GraphQL service’s schema,such as deprecated fields on a type or deprecated enum values."
directive @deprecated("Deprecations include a reason for why it is deprecated, which is formatted using Markdown syntax (as specified by CommonMark)." reason: String = "No longer supported") on FIELD_DEFINITION | ENUM_VALUE

directive @export(as: String) on FIELD

"Directs the executor to include this field or fragment only when the `if` argument is true."
directive @include("Included when true." if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"Directs the executor to skip this field or fragment when the `if` argument is true."
directive @skip("Skipped when true." if: Boolean!) on FIELD | FRAGMENT_SPREAD | INLINE_FRAGMENT

"The `@stream` directive may be provided for a field of `List` type so that the backend can leverage technology such as asynchronous iterators to provide a partial list in the initial response, and additional list items in subsequent responses. `@include` and `@skip` take precedence over `@stream`."
directive @stream("If this argument label has a value other than null, it will be passed on to the result of this stream directive. This label is intended to give client applications a way to identify to which fragment a streamed result belongs to." label: String "The initial elements that shall be send down to the consumer." initialCount: Int! "Streamed when true." if: Boolean!) on FIELD
